### ISSUE TITLE
ccb_explore_adf reports undefined holdout loss if slot-specific features are omitted and --passes > 1

### ISSUE THREAD

**Issue by maxpagels** [VowpalWabbit/vowpal_wabbit#2781]
#### Describe the bug
CCB reports undefined holdout loss if slot-specific features are omitted. Datasets with the following format yields a holdout loss if using more than one pass:

```
ccb shared |User a
ccb action |Action category_trousers price:100
ccb action |Action category_top price:150 size_XS
ccb action |Action category_hat price:150
ccb action |Action category_trousers price:150
ccb action |Action category_top price:200
ccb action |Action category_hat price:200
ccb slot 0:0:0.16 |Slot a
ccb slot 1:0:0.2 |Slot b
```
Whereas datasets with no slot features or a slot namespace with no features reports `average loss = undefined (no holdout)` (note the last two rows in the example):

```
ccb shared |User a
ccb action |Action category_trousers price:100
ccb action |Action category_top price:150 size_XS
ccb action |Action category_hat price:150
ccb action |Action category_trousers price:150
ccb action |Action category_top price:200
ccb action |Action category_hat price:200
ccb slot 0:1:0.16 |Slot
ccb slot 1:0:0.2 |
```

#### To Reproduce

Download the following test files: 
[bar.txt](https://github.com/VowpalWabbit/vowpal_wabbit/files/5831350/bar.txt)
[foo.txt](https://github.com/VowpalWabbit/vowpal_wabbit/files/5831351/foo.txt)

1. Run `vw --ccb_explore_adf -d foo.txt --passes 10 -c -k` and observe holdout loss `average loss = 0.024479 h`
2. Run `vw --ccb_explore_adf -d bar.txt --passes 10 -c -k` and observe holdout loss `average loss = undefined (no holdout)`

#### Expected behavior
Omitting slot features is normal if none are available. It shouldn't affect the possibility to train over several epochs and get a holdout loss. This is useful in settings where we learn on a schedule and could benefit from several passes over data.

Moreover, CCBs should be possible to train without exploration (like `cb_adf` instead of `cb_explore_adf`; CCBs have no `ccb_adf` option, only `ccb_explore_adf`). Reason: offline evaluation of new policies.

#### Observed Behavior
`average loss = undefined (no holdout)` instead of proper holdout loss

#### Environment
VW version: 8.9.0
OS: Mac Big Sur, CLI option

**Comment by olgavrou** [comment]
Hi @maxpagels this should be resolved now in master, but please feel free to open again if the problem persists.
